{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "hourly-jordan",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, copy, json, bz2\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "pandas.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caroline-perry",
   "metadata": {},
   "source": [
    "# Creating the COGUK dataset\n",
    "\n",
    "First, let's load in the COGUK metadata file that is publicly available via\n",
    "\n",
    "`wget https://cog-uk.s3.climb.ac.uk/phylogenetics/latest/cog_metadata.csv\n",
    "`\n",
    "\n",
    "Note that we can infer `cogid` from the `sequence_name` and it is not guaranteed unique, so we shall keep the last."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "sitting-identity",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 506543 rows in the dataset\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>cog_country</th>\n",
       "      <th>adm1</th>\n",
       "      <th>is_pillar_2</th>\n",
       "      <th>sample_date</th>\n",
       "      <th>epi_week</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineages_version</th>\n",
       "      <th>lineage_conflict</th>\n",
       "      <th>lineage_ambiguity_score</th>\n",
       "      <th>scorpio_call</th>\n",
       "      <th>scorpio_support</th>\n",
       "      <th>scorpio_conflict</th>\n",
       "      <th>t1001i</th>\n",
       "      <th>n501y</th>\n",
       "      <th>d614g</th>\n",
       "      <th>del_21765_6</th>\n",
       "      <th>y453f</th>\n",
       "      <th>del_1605_3</th>\n",
       "      <th>a222v</th>\n",
       "      <th>n439k</th>\n",
       "      <th>mutations</th>\n",
       "      <th>p681h</th>\n",
       "      <th>q27stop</th>\n",
       "      <th>e484k</th>\n",
       "      <th>p323l</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cogid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PHEC-L307LF63</th>\n",
       "      <td>Northern_Ireland/PHEC-L307LF63/2021</td>\n",
       "      <td>UK</td>\n",
       "      <td>UK-NIR</td>\n",
       "      <td>N</td>\n",
       "      <td>2021-06-15</td>\n",
       "      <td>77</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>PLEARN-v1.2.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>ref</td>\n",
       "      <td>Y</td>\n",
       "      <td>ref</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>synSNP:G1186C|synSNP:C3037T|orf1ab:A1306S|orf1...</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>E</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHEC-L307LF81</th>\n",
       "      <td>Northern_Ireland/PHEC-L307LF81/2021</td>\n",
       "      <td>UK</td>\n",
       "      <td>UK-NIR</td>\n",
       "      <td>N</td>\n",
       "      <td>2021-06-15</td>\n",
       "      <td>77</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>PLEARN-v1.2.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>ref</td>\n",
       "      <td>Y</td>\n",
       "      <td>ref</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>synSNP:C3037T|orf1ab:A1306S|orf1ab:P2046L|orf1...</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>E</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PHEC-L307LF45</th>\n",
       "      <td>Northern_Ireland/PHEC-L307LF45/2021</td>\n",
       "      <td>UK</td>\n",
       "      <td>UK-NIR</td>\n",
       "      <td>N</td>\n",
       "      <td>2021-06-15</td>\n",
       "      <td>77</td>\n",
       "      <td>B.1.617.2</td>\n",
       "      <td>PLEARN-v1.2.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Delta (B.1.617.2-like)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>ref</td>\n",
       "      <td>Y</td>\n",
       "      <td>ref</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>orf1ab:P309L|synSNP:C1267T|synSNP:C3037T|orf1a...</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>E</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     sequence_name cog_country    adm1  \\\n",
       "cogid                                                                    \n",
       "PHEC-L307LF63  Northern_Ireland/PHEC-L307LF63/2021          UK  UK-NIR   \n",
       "PHEC-L307LF81  Northern_Ireland/PHEC-L307LF81/2021          UK  UK-NIR   \n",
       "PHEC-L307LF45  Northern_Ireland/PHEC-L307LF45/2021          UK  UK-NIR   \n",
       "\n",
       "              is_pillar_2 sample_date  epi_week    lineage lineages_version  \\\n",
       "cogid                                                                         \n",
       "PHEC-L307LF63           N  2021-06-15        77  B.1.617.2   PLEARN-v1.2.13   \n",
       "PHEC-L307LF81           N  2021-06-15        77  B.1.617.2   PLEARN-v1.2.13   \n",
       "PHEC-L307LF45           N  2021-06-15        77  B.1.617.2   PLEARN-v1.2.13   \n",
       "\n",
       "               lineage_conflict  lineage_ambiguity_score  \\\n",
       "cogid                                                      \n",
       "PHEC-L307LF63               NaN                      NaN   \n",
       "PHEC-L307LF81               NaN                      NaN   \n",
       "PHEC-L307LF45               NaN                      NaN   \n",
       "\n",
       "                         scorpio_call  scorpio_support  scorpio_conflict  \\\n",
       "cogid                                                                      \n",
       "PHEC-L307LF63  Delta (B.1.617.2-like)              1.0               0.0   \n",
       "PHEC-L307LF81  Delta (B.1.617.2-like)              1.0               0.0   \n",
       "PHEC-L307LF45  Delta (B.1.617.2-like)              1.0               0.0   \n",
       "\n",
       "              t1001i n501y d614g del_21765_6 y453f del_1605_3 a222v n439k  \\\n",
       "cogid                                                                       \n",
       "PHEC-L307LF63      T     N     G         ref     Y        ref     A     N   \n",
       "PHEC-L307LF81      T     N     G         ref     Y        ref     A     N   \n",
       "PHEC-L307LF45      T     N     G         ref     Y        ref     A     N   \n",
       "\n",
       "                                                       mutations p681h  \\\n",
       "cogid                                                                    \n",
       "PHEC-L307LF63  synSNP:G1186C|synSNP:C3037T|orf1ab:A1306S|orf1...     R   \n",
       "PHEC-L307LF81  synSNP:C3037T|orf1ab:A1306S|orf1ab:P2046L|orf1...     R   \n",
       "PHEC-L307LF45  orf1ab:P309L|synSNP:C1267T|synSNP:C3037T|orf1a...     R   \n",
       "\n",
       "              q27stop e484k p323l  \n",
       "cogid                              \n",
       "PHEC-L307LF63       Q     E     L  \n",
       "PHEC-L307LF81       Q     E     L  \n",
       "PHEC-L307LF45       Q     E     L  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_metadata=pandas.read_csv('cog_metadata-20210624.csv.gz')\n",
    "\n",
    "def split_sequence_name(row):\n",
    "    cols=row['sequence_name'].split('/')\n",
    "    if len(cols)==3:\n",
    "        return(cols[1])\n",
    "    else:\n",
    "        return(None)\n",
    "\n",
    "# rename the country to avoid a conflict with the ENA dataset later\n",
    "cog_metadata.rename(columns={'country':'cog_country'},inplace=True)\n",
    "\n",
    "# infer the cogid\n",
    "cog_metadata['cogid']=cog_metadata.apply(split_sequence_name,axis=1)        \n",
    "\n",
    "# set as the index and de-duplicate\n",
    "cog_metadata.set_index('cogid',inplace=True,verify_integrity=False)\n",
    "cog_metadata=cog_metadata.loc[~cog_metadata.index.duplicated(keep='first')]\n",
    "cog_metadata.reset_index(inplace=True)\n",
    "cog_metadata.set_index('cogid',inplace=True,verify_integrity=True)\n",
    "\n",
    "print(\"There are \"+str(len(cog_metadata))+\" rows in the dataset\")\n",
    "\n",
    "cog_metadata[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "experimental-affiliate",
   "metadata": {},
   "source": [
    "But we don't have the ENA `run_accession` which we need to link to the FASTQ files in object store, nor do we know what the sequencing platform was, which we need to choose the right flow in SP3, so let's load in a list from the ENA and again we can infer `cogid` from the `submitted_ftp` file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "executed-heritage",
   "metadata": {},
   "source": [
    "To get the latest TSV issue the below in a terminal; be warned that unless you subset the fields it will be >2GB; either way best to compress.\n",
    "\n",
    "`curl \"https://www.ebi.ac.uk/ena/portal/api/filereport?accession=PRJEB37886&result=read_run&fields=study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,instrument_platform,country,submitted_ftp,read_count,base_count&limit=1000000&format=tsv&download=true\" > cog-ena-20210624.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fancy-colonial",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>secondary_study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>nominal_length</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>read_count</th>\n",
       "      <th>base_count</th>\n",
       "      <th>center_name</th>\n",
       "      <th>first_public</th>\n",
       "      <th>last_updated</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>fastq_bytes</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>fastq_aspera</th>\n",
       "      <th>fastq_galaxy</th>\n",
       "      <th>submitted_bytes</th>\n",
       "      <th>submitted_md5</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>submitted_aspera</th>\n",
       "      <th>submitted_galaxy</th>\n",
       "      <th>submitted_format</th>\n",
       "      <th>sra_bytes</th>\n",
       "      <th>sra_md5</th>\n",
       "      <th>sra_ftp</th>\n",
       "      <th>sra_aspera</th>\n",
       "      <th>sra_galaxy</th>\n",
       "      <th>cram_index_ftp</th>\n",
       "      <th>cram_index_aspera</th>\n",
       "      <th>cram_index_galaxy</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>nominal_sdev</th>\n",
       "      <th>first_created</th>\n",
       "      <th>sample_description</th>\n",
       "      <th>parent_study</th>\n",
       "      <th>sample_accession.1</th>\n",
       "      <th>accession</th>\n",
       "      <th>secondary_sample_accession.1</th>\n",
       "      <th>bio_material</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>collected_by</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>country</th>\n",
       "      <th>cultivar</th>\n",
       "      <th>culture_collection</th>\n",
       "      <th>description</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>ecotype</th>\n",
       "      <th>environmental_sample</th>\n",
       "      <th>first_public.1</th>\n",
       "      <th>germline</th>\n",
       "      <th>identified_by</th>\n",
       "      <th>isolate</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>location</th>\n",
       "      <th>mating_type</th>\n",
       "      <th>serotype</th>\n",
       "      <th>serovar</th>\n",
       "      <th>sex</th>\n",
       "      <th>submitted_sex</th>\n",
       "      <th>specimen_voucher</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>sub_strain</th>\n",
       "      <th>tissue_lib</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>variety</th>\n",
       "      <th>tax_id.1</th>\n",
       "      <th>scientific_name.1</th>\n",
       "      <th>sample_alias.1</th>\n",
       "      <th>checklist</th>\n",
       "      <th>center_name.1</th>\n",
       "      <th>depth</th>\n",
       "      <th>elevation</th>\n",
       "      <th>altitude</th>\n",
       "      <th>environment_biome</th>\n",
       "      <th>environment_feature</th>\n",
       "      <th>environment_material</th>\n",
       "      <th>temperature</th>\n",
       "      <th>salinity</th>\n",
       "      <th>sampling_campaign</th>\n",
       "      <th>sampling_site</th>\n",
       "      <th>sampling_platform</th>\n",
       "      <th>protocol_label</th>\n",
       "      <th>project_name</th>\n",
       "      <th>host</th>\n",
       "      <th>host_tax_id</th>\n",
       "      <th>host_status</th>\n",
       "      <th>host_sex</th>\n",
       "      <th>submitted_host_sex</th>\n",
       "      <th>host_body_site</th>\n",
       "      <th>host_gravidity</th>\n",
       "      <th>host_phenotype</th>\n",
       "      <th>host_genotype</th>\n",
       "      <th>host_growth_conditions</th>\n",
       "      <th>environmental_package</th>\n",
       "      <th>investigation_type</th>\n",
       "      <th>experimental_factor</th>\n",
       "      <th>sample_collection</th>\n",
       "      <th>sequencing_method</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>ph</th>\n",
       "      <th>broker_name.1</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>sample_material</th>\n",
       "      <th>taxonomic_identity_marker</th>\n",
       "      <th>assembly_quality</th>\n",
       "      <th>assembly_software</th>\n",
       "      <th>taxonomic_classification</th>\n",
       "      <th>completeness_score</th>\n",
       "      <th>contamination_score</th>\n",
       "      <th>binning_software</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>sample_capture_status</th>\n",
       "      <th>collection_date_submitted</th>\n",
       "      <th>submission_tool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJEB37886</td>\n",
       "      <td>ERP121228</td>\n",
       "      <td>SAMEA6791711</td>\n",
       "      <td>ERS4519383</td>\n",
       "      <td>ERX4078237</td>\n",
       "      <td>ERR4082025</td>\n",
       "      <td>ERA2524107</td>\n",
       "      <td>2697049</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>OXFORD_NANOPORE</td>\n",
       "      <td>MinION</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>211388</td>\n",
       "      <td>103672824</td>\n",
       "      <td>University of Cambridge</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>MinION sequencing; COG-UK/CAMB-7374F/CAMB:2020...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) consortium</td>\n",
       "      <td>ena-STUDY-COVID-19 Genomics UK Consortium-18-0...</td>\n",
       "      <td>COG-UK/CAMB-7374F/CAMB:20200404_1455_MN30591_F...</td>\n",
       "      <td>COG-UK/CAMB-7374F/CAMB:20200404_1455_MN30591_F...</td>\n",
       "      <td>91988098</td>\n",
       "      <td>28d6f7213833a6e5025fb4d42767e011</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/005/ERR408...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/fastq/ERR408/005/ERR4...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/005/ERR408...</td>\n",
       "      <td>108722728</td>\n",
       "      <td>18f53e799c25060ad74ae240c253f3a8</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/run/ERR408/ERR4082025...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...</td>\n",
       "      <td>BAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COG-UK/CAMB-7374F</td>\n",
       "      <td>COVID-19 Genomics UK Consortium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>COG-UK/CAMB-7374F</td>\n",
       "      <td>PRJEB39908</td>\n",
       "      <td>SAMEA6791711</td>\n",
       "      <td>SAMEA6791711</td>\n",
       "      <td>ERS4519383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MinION sequencing; COG-UK/CAMB-7374F/CAMB:2020...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2697049</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>COG-UK/CAMB-7374F</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>University of Cambridge</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COVID-19 Genomics UK Consortium</td>\n",
       "      <td>COG-UK/CAMB-7374F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJEB37886</td>\n",
       "      <td>ERP121228</td>\n",
       "      <td>SAMEA6791712</td>\n",
       "      <td>ERS4519384</td>\n",
       "      <td>ERX4078238</td>\n",
       "      <td>ERR4082026</td>\n",
       "      <td>ERA2524109</td>\n",
       "      <td>2697049</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>OXFORD_NANOPORE</td>\n",
       "      <td>MinION</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>VIRAL RNA</td>\n",
       "      <td>PCR</td>\n",
       "      <td>427553</td>\n",
       "      <td>210103657</td>\n",
       "      <td>University of Cambridge</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>MinION sequencing; COG-UK/CAMB-7375E/CAMB:2020...</td>\n",
       "      <td>COVID-19 Genomics UK (COG-UK) consortium</td>\n",
       "      <td>ena-STUDY-COVID-19 Genomics UK Consortium-18-0...</td>\n",
       "      <td>COG-UK/CAMB-7375E/CAMB:20200404_1455_MN30591_F...</td>\n",
       "      <td>COG-UK/CAMB-7375E/CAMB:20200404_1455_MN30591_F...</td>\n",
       "      <td>186032909</td>\n",
       "      <td>811cf092a7ea2ac765211fc38e6b5c74</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/006/ERR408...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/fastq/ERR408/006/ERR4...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/006/ERR408...</td>\n",
       "      <td>220151623</td>\n",
       "      <td>39d9f40115d2a4b63503efc40ed12906</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...</td>\n",
       "      <td>fasp.sra.ebi.ac.uk:/vol1/run/ERR408/ERR4082026...</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...</td>\n",
       "      <td>BAM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COG-UK/CAMB-7375E</td>\n",
       "      <td>COVID-19 Genomics UK Consortium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>COG-UK/CAMB-7375E</td>\n",
       "      <td>PRJEB39908</td>\n",
       "      <td>SAMEA6791712</td>\n",
       "      <td>SAMEA6791712</td>\n",
       "      <td>ERS4519384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MinION sequencing; COG-UK/CAMB-7375E/CAMB:2020...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>2020-04-29</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2697049</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>COG-UK/CAMB-7375E</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>University of Cambridge</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COVID-19 Genomics UK Consortium</td>\n",
       "      <td>COG-UK/CAMB-7375E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession secondary_study_accession sample_accession  \\\n",
       "0      PRJEB37886                 ERP121228     SAMEA6791711   \n",
       "1      PRJEB37886                 ERP121228     SAMEA6791712   \n",
       "\n",
       "  secondary_sample_accession experiment_accession run_accession  \\\n",
       "0                 ERS4519383           ERX4078237    ERR4082025   \n",
       "1                 ERS4519384           ERX4078238    ERR4082026   \n",
       "\n",
       "  submission_accession   tax_id  \\\n",
       "0           ERA2524107  2697049   \n",
       "1           ERA2524109  2697049   \n",
       "\n",
       "                                   scientific_name instrument_platform  \\\n",
       "0  Severe acute respiratory syndrome coronavirus 2     OXFORD_NANOPORE   \n",
       "1  Severe acute respiratory syndrome coronavirus 2     OXFORD_NANOPORE   \n",
       "\n",
       "  instrument_model  library_name library_layout  nominal_length  \\\n",
       "0           MinION           NaN         SINGLE             NaN   \n",
       "1           MinION           NaN         SINGLE             NaN   \n",
       "\n",
       "  library_strategy library_source library_selection  read_count  base_count  \\\n",
       "0         AMPLICON      VIRAL RNA               PCR      211388   103672824   \n",
       "1         AMPLICON      VIRAL RNA               PCR      427553   210103657   \n",
       "\n",
       "               center_name first_public last_updated  \\\n",
       "0  University of Cambridge   2020-04-29   2020-04-29   \n",
       "1  University of Cambridge   2020-04-29   2020-04-29   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  MinION sequencing; COG-UK/CAMB-7374F/CAMB:2020...   \n",
       "1  MinION sequencing; COG-UK/CAMB-7375E/CAMB:2020...   \n",
       "\n",
       "                                study_title  \\\n",
       "0  COVID-19 Genomics UK (COG-UK) consortium   \n",
       "1  COVID-19 Genomics UK (COG-UK) consortium   \n",
       "\n",
       "                                         study_alias  \\\n",
       "0  ena-STUDY-COVID-19 Genomics UK Consortium-18-0...   \n",
       "1  ena-STUDY-COVID-19 Genomics UK Consortium-18-0...   \n",
       "\n",
       "                                    experiment_alias  \\\n",
       "0  COG-UK/CAMB-7374F/CAMB:20200404_1455_MN30591_F...   \n",
       "1  COG-UK/CAMB-7375E/CAMB:20200404_1455_MN30591_F...   \n",
       "\n",
       "                                           run_alias  fastq_bytes  \\\n",
       "0  COG-UK/CAMB-7374F/CAMB:20200404_1455_MN30591_F...     91988098   \n",
       "1  COG-UK/CAMB-7375E/CAMB:20200404_1455_MN30591_F...    186032909   \n",
       "\n",
       "                          fastq_md5  \\\n",
       "0  28d6f7213833a6e5025fb4d42767e011   \n",
       "1  811cf092a7ea2ac765211fc38e6b5c74   \n",
       "\n",
       "                                           fastq_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/005/ERR408...   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/006/ERR408...   \n",
       "\n",
       "                                        fastq_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/fastq/ERR408/005/ERR4...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/fastq/ERR408/006/ERR4...   \n",
       "\n",
       "                                        fastq_galaxy  submitted_bytes  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/005/ERR408...        108722728   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR408/006/ERR408...        220151623   \n",
       "\n",
       "                      submitted_md5  \\\n",
       "0  18f53e799c25060ad74ae240c253f3a8   \n",
       "1  39d9f40115d2a4b63503efc40ed12906   \n",
       "\n",
       "                                       submitted_ftp  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...   \n",
       "\n",
       "                                    submitted_aspera  \\\n",
       "0  fasp.sra.ebi.ac.uk:/vol1/run/ERR408/ERR4082025...   \n",
       "1  fasp.sra.ebi.ac.uk:/vol1/run/ERR408/ERR4082026...   \n",
       "\n",
       "                                    submitted_galaxy submitted_format  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...              BAM   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...              BAM   \n",
       "\n",
       "   sra_bytes  sra_md5  sra_ftp  sra_aspera  sra_galaxy  cram_index_ftp  \\\n",
       "0        NaN      NaN      NaN         NaN         NaN             NaN   \n",
       "1        NaN      NaN      NaN         NaN         NaN             NaN   \n",
       "\n",
       "   cram_index_aspera  cram_index_galaxy       sample_alias  \\\n",
       "0                NaN                NaN  COG-UK/CAMB-7374F   \n",
       "1                NaN                NaN  COG-UK/CAMB-7375E   \n",
       "\n",
       "                       broker_name  nominal_sdev first_created  \\\n",
       "0  COVID-19 Genomics UK Consortium           NaN    2020-04-29   \n",
       "1  COVID-19 Genomics UK Consortium           NaN    2020-04-29   \n",
       "\n",
       "  sample_description parent_study sample_accession.1     accession  \\\n",
       "0  COG-UK/CAMB-7374F   PRJEB39908       SAMEA6791711  SAMEA6791711   \n",
       "1  COG-UK/CAMB-7375E   PRJEB39908       SAMEA6791712  SAMEA6791712   \n",
       "\n",
       "  secondary_sample_accession.1  bio_material  cell_line  cell_type  \\\n",
       "0                   ERS4519383           NaN        NaN        NaN   \n",
       "1                   ERS4519384           NaN        NaN        NaN   \n",
       "\n",
       "   collected_by  collection_date  country  cultivar  culture_collection  \\\n",
       "0           NaN              NaN      NaN       NaN                 NaN   \n",
       "1           NaN              NaN      NaN       NaN                 NaN   \n",
       "\n",
       "                                         description  dev_stage  ecotype  \\\n",
       "0  MinION sequencing; COG-UK/CAMB-7374F/CAMB:2020...        NaN      NaN   \n",
       "1  MinION sequencing; COG-UK/CAMB-7375E/CAMB:2020...        NaN      NaN   \n",
       "\n",
       "   environmental_sample first_public.1  germline  identified_by  isolate  \\\n",
       "0                 False     2020-04-29     False            NaN      NaN   \n",
       "1                 False     2020-04-29     False            NaN      NaN   \n",
       "\n",
       "   isolation_source  location  mating_type  serotype  serovar  sex  \\\n",
       "0               NaN       NaN          NaN       NaN      NaN  NaN   \n",
       "1               NaN       NaN          NaN       NaN      NaN  NaN   \n",
       "\n",
       "   submitted_sex  specimen_voucher  strain  sub_species  sub_strain  \\\n",
       "0            NaN               NaN     NaN          NaN         NaN   \n",
       "1            NaN               NaN     NaN          NaN         NaN   \n",
       "\n",
       "   tissue_lib  tissue_type  variety  tax_id.1  \\\n",
       "0         NaN          NaN      NaN   2697049   \n",
       "1         NaN          NaN      NaN   2697049   \n",
       "\n",
       "                                 scientific_name.1     sample_alias.1  \\\n",
       "0  Severe acute respiratory syndrome coronavirus 2  COG-UK/CAMB-7374F   \n",
       "1  Severe acute respiratory syndrome coronavirus 2  COG-UK/CAMB-7375E   \n",
       "\n",
       "   checklist            center_name.1  depth  elevation  altitude  \\\n",
       "0  ERC000011  University of Cambridge    NaN        NaN       NaN   \n",
       "1  ERC000011  University of Cambridge    NaN        NaN       NaN   \n",
       "\n",
       "   environment_biome  environment_feature  environment_material  temperature  \\\n",
       "0                NaN                  NaN                   NaN          NaN   \n",
       "1                NaN                  NaN                   NaN          NaN   \n",
       "\n",
       "   salinity  sampling_campaign  sampling_site  sampling_platform  \\\n",
       "0       NaN                NaN            NaN                NaN   \n",
       "1       NaN                NaN            NaN                NaN   \n",
       "\n",
       "   protocol_label  project_name  host  host_tax_id  host_status  host_sex  \\\n",
       "0             NaN           NaN   NaN          NaN          NaN       NaN   \n",
       "1             NaN           NaN   NaN          NaN          NaN       NaN   \n",
       "\n",
       "   submitted_host_sex  host_body_site  host_gravidity  host_phenotype  \\\n",
       "0                 NaN             NaN             NaN             NaN   \n",
       "1                 NaN             NaN             NaN             NaN   \n",
       "\n",
       "   host_genotype  host_growth_conditions  environmental_package  \\\n",
       "0            NaN                     NaN                    NaN   \n",
       "1            NaN                     NaN                    NaN   \n",
       "\n",
       "   investigation_type  experimental_factor  sample_collection  \\\n",
       "0                 NaN                  NaN                NaN   \n",
       "1                 NaN                  NaN                NaN   \n",
       "\n",
       "   sequencing_method  target_gene  ph                    broker_name.1  \\\n",
       "0                NaN          NaN NaN  COVID-19 Genomics UK Consortium   \n",
       "1                NaN          NaN NaN  COVID-19 Genomics UK Consortium   \n",
       "\n",
       "        sample_title  sample_material  taxonomic_identity_marker  \\\n",
       "0  COG-UK/CAMB-7374F              NaN                        NaN   \n",
       "1  COG-UK/CAMB-7375E              NaN                        NaN   \n",
       "\n",
       "   assembly_quality  assembly_software  taxonomic_classification  \\\n",
       "0               NaN                NaN                       NaN   \n",
       "1               NaN                NaN                       NaN   \n",
       "\n",
       "   completeness_score  contamination_score  binning_software  lat  lon  \\\n",
       "0                 NaN                  NaN               NaN  NaN  NaN   \n",
       "1                 NaN                  NaN               NaN  NaN  NaN   \n",
       "\n",
       "   sample_capture_status  collection_date_submitted  submission_tool  \n",
       "0                    NaN                        NaN              NaN  \n",
       "1                    NaN                        NaN              NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ena_sample=pandas.read_csv('ena_samples_all_fields.tsv',sep='\\t')\n",
    "ena_sample[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "considered-pressure",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 496763 rows in the dataset\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>instrument_platform</th>\n",
       "      <th>ena_country</th>\n",
       "      <th>submitted_ftp</th>\n",
       "      <th>read_count</th>\n",
       "      <th>base_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cogid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAMB-7374F</th>\n",
       "      <td>PRJEB37886</td>\n",
       "      <td>SAMEA6791711</td>\n",
       "      <td>ERS4519383</td>\n",
       "      <td>ERX4078237</td>\n",
       "      <td>ERR4082025</td>\n",
       "      <td>OXFORD_NANOPORE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...</td>\n",
       "      <td>211388.0</td>\n",
       "      <td>103672824.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAMB-7375E</th>\n",
       "      <td>PRJEB37886</td>\n",
       "      <td>SAMEA6791712</td>\n",
       "      <td>ERS4519384</td>\n",
       "      <td>ERX4078238</td>\n",
       "      <td>ERR4082026</td>\n",
       "      <td>OXFORD_NANOPORE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...</td>\n",
       "      <td>427553.0</td>\n",
       "      <td>210103657.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAMB-7376D</th>\n",
       "      <td>PRJEB37886</td>\n",
       "      <td>SAMEA6791713</td>\n",
       "      <td>ERS4519385</td>\n",
       "      <td>ERX4078239</td>\n",
       "      <td>ERR4082027</td>\n",
       "      <td>OXFORD_NANOPORE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082027/C...</td>\n",
       "      <td>374754.0</td>\n",
       "      <td>183823695.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           study_accession sample_accession secondary_sample_accession  \\\n",
       "cogid                                                                    \n",
       "CAMB-7374F      PRJEB37886     SAMEA6791711                 ERS4519383   \n",
       "CAMB-7375E      PRJEB37886     SAMEA6791712                 ERS4519384   \n",
       "CAMB-7376D      PRJEB37886     SAMEA6791713                 ERS4519385   \n",
       "\n",
       "           experiment_accession run_accession instrument_platform ena_country  \\\n",
       "cogid                                                                           \n",
       "CAMB-7374F           ERX4078237    ERR4082025     OXFORD_NANOPORE         NaN   \n",
       "CAMB-7375E           ERX4078238    ERR4082026     OXFORD_NANOPORE         NaN   \n",
       "CAMB-7376D           ERX4078239    ERR4082027     OXFORD_NANOPORE         NaN   \n",
       "\n",
       "                                                submitted_ftp  read_count  \\\n",
       "cogid                                                                       \n",
       "CAMB-7374F  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082025/C...    211388.0   \n",
       "CAMB-7375E  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082026/C...    427553.0   \n",
       "CAMB-7376D  ftp.sra.ebi.ac.uk/vol1/run/ERR408/ERR4082027/C...    374754.0   \n",
       "\n",
       "             base_count  \n",
       "cogid                    \n",
       "CAMB-7374F  103672824.0  \n",
       "CAMB-7375E  210103657.0  \n",
       "CAMB-7376D  183823695.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog_ena=pandas.read_csv('cog_ena-20210624.tsv.gz',sep='\\t')\n",
    "\n",
    "def create_cogid(row):\n",
    "    if isinstance(row['submitted_ftp'],str):\n",
    "        return(row['submitted_ftp'].split('/')[-1].split('.')[0])\n",
    "    else:\n",
    "        return(None)\n",
    "\n",
    "# rename the country to avoid a conflict with the metadata dataset later\n",
    "cog_ena.rename(columns={'country':'ena_country'},inplace=True)\n",
    "\n",
    "# filter down to Illumina and ONT only (excludes Ion Torrent)\n",
    "cog_ena=cog_ena.loc[cog_ena.instrument_platform.isin(['ILLUMINA','OXFORD_NANOPORE'])]\n",
    "\n",
    "# infer the cogid\n",
    "cog_ena['cogid']=cog_ena.apply(create_cogid,axis=1)\n",
    "\n",
    "# remove any samples with no cogid\n",
    "cog_ena=cog_ena.loc[cog_ena.cogid.notna()]\n",
    "\n",
    "# set as the index and de-duplicate\n",
    "cog_ena.set_index('cogid',inplace=True,verify_integrity=False)\n",
    "cog_ena=cog_ena.loc[~cog_ena.index.duplicated(keep='first')]\n",
    "cog_ena.reset_index(inplace=True)\n",
    "cog_ena.set_index('cogid',inplace=True,verify_integrity=True)\n",
    "\n",
    "print(\"There are \"+str(len(cog_ena))+\" rows in the dataset\")\n",
    "\n",
    "cog_ena[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ahead-stress",
   "metadata": {},
   "source": [
    "Now let's do an inner join so every row has a unique `cogid` and an ENA `run_accession`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "dated-sunrise",
   "metadata": {},
   "outputs": [],
   "source": [
    "cog=cog_metadata.join(cog_ena,how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "handed-spotlight",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This gives us 455352 COGUK samples with an ENA accession number\n",
      "There are 0 samples with no PANGO lineage and 169437 samples with no SCORPIO lineage\n"
     ]
    }
   ],
   "source": [
    "print(\"This gives us \"+str(len(cog))+\" COGUK samples with an ENA accession number\")\n",
    "\n",
    "print(\"There are \" +str(len(cog.loc[cog.lineage.isna()]))+\" samples with no PANGO lineage and \" +str(len(cog.loc[cog.scorpio_call.isna()]))+ \" samples with no SCORPIO lineage\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "naval-mongolia",
   "metadata": {},
   "source": [
    "Now let's match to what is in the `ena` bucket on object store so we can make sure we are only asking for run accessions that have fastq files in the bucket."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "hairy-smooth",
   "metadata": {},
   "outputs": [],
   "source": [
    "with bz2.open('oci-ena-bucket-20210625.json.bz2') as jsonFile:\n",
    "    jsonObject = json.load(jsonFile)\n",
    "    jsonFile.close()\n",
    "    \n",
    "jsonPayload=jsonObject['data']    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "hourly-edwards",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 764990/764990 [00:00<00:00, 812165.37it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 308136 unique ENA run accessions stored in the ena bucket in object store\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "ena_accessions=set()\n",
    "\n",
    "for i in tqdm(jsonPayload):\n",
    "    \n",
    "    fastaFile=i['name']\n",
    "    \n",
    "    if \"_\" in fastaFile:\n",
    "        era=fastaFile.split('_')[0]\n",
    "    else:\n",
    "        era=fastaFile.split('.')[0]\n",
    "    if era not in ena_accessions:\n",
    "        ena_accessions.add(era)\n",
    "\n",
    "print(\"There are \"+str(len(ena_accessions))+\" unique ENA run accessions stored in the ena bucket in object store\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "stock-lewis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     228773\n",
       "False    226579\n",
       "Name: in_ena_bucket, dtype: int64"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cog['in_ena_bucket']=cog.run_accession.isin(ena_accessions)\n",
    "cog.in_ena_bucket.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "horizontal-mapping",
   "metadata": {},
   "outputs": [],
   "source": [
    "cog.to_csv('cog_dataset-20210624.csv.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "qualified-waste",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
